File: src/main/java/com/raiderrobotix/_2020/subsystems/DriveBase.java
package com.raiderrobotix._2020.subsystems;

import java.util.function.Supplier;

import com.kauailabs.navx.frc.AHRS;
import com.raiderrobotix._2020.commands.Teleop;
import com.revrobotics.CANEncoder;
import com.revrobotics.CANSparkMax;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;

import edu.wpi.first.wpilibj.AnalogInput;
import edu.wpi.first.wpilibj.SerialPort.Port;
import edu.wpi.first.wpilibj.command.Subsystem;
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class DriveBase extends Subsystem {

	public static final DriveBase drives = new DriveBase();

	private static final int LEFT_FRONT_DRIVE_CAN_ID = 2;
	private static final int LEFT_BACK_DRIVE_CAN_ID = 1;
	private static final int RIGHT_FRONT_DRIVE_CAN_ID = 3;
	private static final int RIGHT_BACK_DRIVE_CAN_ID = 4;

	private static final CANSparkMax leftFrontSpark = new CANSparkMax(LEFT_FRONT_DRIVE_CAN_ID, MotorType.kBrushless);
	private static final CANSparkMax leftBackSpark = new CANSparkMax(LEFT_BACK_DRIVE_CAN_ID, MotorType.kBrushless);
	private static final CANSparkMax rightFrontSpark = new CANSparkMax(RIGHT_FRONT_DRIVE_CAN_ID, MotorType.kBrushless);
	private static final CANSparkMax rightBackSpark = new CANSparkMax(RIGHT_BACK_DRIVE_CAN_ID, MotorType.kBrushless);

	private static final double TIRE_CIRCUMFERENCE = 28.375; // TODO

	private static final double GEAR_RATIO = 0.047619; // 0.0714286; // TODO
	private static final double INCHES_PER_REVOLUTION = GEAR_RATIO * TIRE_CIRCUMFERENCE;

	private static final boolean RIGHT_DRIVE_MOTORS_INVERTED = true; // TODO
	private static final boolean LEFT_DRIVE_MOTORS_INVERTED = false; // TODO

	private static final CANEncoder leftEncoder = leftFrontSpark.getEncoder();
	private static final CANEncoder rightEncoder = rightFrontSpark.getEncoder();

	private static final AnalogInput ultrasonic = new AnalogInput(0);

	private static final AHRS navX = new AHRS(Port.kUSB1);

	private DriveBase() {
		leftFrontSpark.setInverted(LEFT_DRIVE_MOTORS_INVERTED);
		rightFrontSpark.setInverted(RIGHT_DRIVE_MOTORS_INVERTED);

		leftBackSpark.follow(leftFrontSpark);
		rightBackSpark.follow(rightFrontSpark);

		leftEncoder.setPositionConversionFactor(INCHES_PER_REVOLUTION);
		rightEncoder.setPositionConversionFactor(INCHES_PER_REVOLUTION);

		SmartDashboard.putData(this);
	}

	public void setSpeed(double speed) {
		setSpeed(speed, speed);
	}

	public void setSpeed(double leftSpeed, double rightSpeed) {
		leftFrontSpark.set(leftSpeed);
		rightFrontSpark.set(rightSpeed);
	}

	public double getAverageDistance() {
		return (getLeftDistance() + getRightDistance()) / 2.0;
	}

	public double getLeftDistance() {
		return leftEncoder.getPosition();
	}

	public double getRightDistance() {
		return rightEncoder.getPosition();
	}

	public void resetEncoders() {
		leftEncoder.setPosition(0);
		rightEncoder.setPosition(0);
	}

	public double getUltrasonicDistance() {
		return ultrasonic.getVoltage();
	}

	public double getGyroAngle() {
		return navX.getAngle();
	}

	public void resetGyro() {
		navX.reset();
	}

	@Override
	protected void initDefaultCommand() {
		setDefaultCommand(new Teleop());
	}

	@Override
	public void initSendable(SendableBuilder builder) {
		builder.setSmartDashboardType("Drives");
		builder.setActuator(true);
		builder.setSafeState(() -> this.setSpeed(0));
		builder.addDoubleProperty("leftDistance", this::getLeftDistance, null);
		builder.addDoubleProperty("rightDistance", this::getRightDistance, null);
		final Supplier<String[]> firmwareSupplier = () -> new String[] {
				"Left Front " + leftFrontSpark.getFirmwareString(), "Left Back " + leftBackSpark.getFirmwareString(),
				"Right Front" + rightFrontSpark.getFirmwareString(),
				"Right Back " + rightBackSpark.getFirmwareString() };

		builder.addStringArrayProperty("firmware", firmwareSupplier, null);
		builder.addDoubleProperty("Gyro", this::getGyroAngle, null);

	}

}

File: src/main/java/com/raiderrobotix/_2020/subsystems/DriveBase.kt
package com.raiderrobotix._2020.subsystems

import com.kauailabs.navx.frc.AHRS
import com.raiderrobotix._2020.commands.Teleop
import com.revrobotics.CANSparkMax
import com.revrobotics.CANSparkMaxLowLevel.MotorType
import edu.wpi.first.wpilibj.AnalogInput
import edu.wpi.first.wpilibj.PIDOutput
import edu.wpi.first.wpilibj.SerialPort.Port
import edu.wpi.first.wpilibj.command.Subsystem
import edu.wpi.first.wpilibj.smartdashboard.SendableBuilder
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard

object DriveBase : Subsystem(), PIDOutput {
	
	private const val LEFT_FRONT_DRIVE_CAN_ID = 2
	private const val LEFT_BACK_DRIVE_CAN_ID = 1
	private const val RIGHT_FRONT_DRIVE_CAN_ID = 3
	private const val RIGHT_BACK_DRIVE_CAN_ID = 4
	
	private val leftFrontSpark = CANSparkMax(LEFT_FRONT_DRIVE_CAN_ID, MotorType.kBrushless)
	private val leftBackSpark = CANSparkMax(LEFT_BACK_DRIVE_CAN_ID, MotorType.kBrushless)
	private val rightFrontSpark = CANSparkMax(RIGHT_FRONT_DRIVE_CAN_ID, MotorType.kBrushless)
	private val rightBackSpark = CANSparkMax(RIGHT_BACK_DRIVE_CAN_ID, MotorType.kBrushless)
	
	private const val TIRE_CIRCUMFERENCE = 28.375 // TODO
	private const val GEAR_RATIO = 0.047619 // 0.0714286; // TODO
	private const val INCHES_PER_REVOLUTION = GEAR_RATIO * TIRE_CIRCUMFERENCE
	private const val RIGHT_DRIVE_MOTORS_INVERTED = true // TODO
	private const val LEFT_DRIVE_MOTORS_INVERTED = false // TODO
	
	private val leftEncoder = leftFrontSpark.encoder
	private val rightEncoder = rightFrontSpark.encoder
	private val ultrasonic = AnalogInput(0)
	private val navX = AHRS(Port.kUSB1)
	
	init {
		leftFrontSpark.inverted = LEFT_DRIVE_MOTORS_INVERTED
		rightFrontSpark.inverted = RIGHT_DRIVE_MOTORS_INVERTED
		leftBackSpark.follow(leftFrontSpark)
		rightBackSpark.follow(rightFrontSpark)
		leftEncoder.positionConversionFactor = INCHES_PER_REVOLUTION
		rightEncoder.positionConversionFactor = INCHES_PER_REVOLUTION
		navX.subsystem = "DriveBase"
		SmartDashboard.putData(this)
	}
	
	val averageDistance: Double
		get() = (leftDistance + rightDistance) / 2.0
	
	private val leftDistance: Double
		get() = leftEncoder.position
	
	private val rightDistance: Double
		get() = rightEncoder.position
	
	val ultrasonicDistance: Double
		get() = ultrasonic.voltage
	
	val gyroAngle: Double
		get() = navX.angle
	
	var speed: Double = 0.0
		set(it) {
			setSpeed(it, it)
			field = it
		}
	
	fun setSpeed(leftSpeed: Double, rightSpeed: Double) {
		leftFrontSpark.set(leftSpeed)
		rightFrontSpark.set(rightSpeed)
	}
	
	fun resetEncoders() {
		leftEncoder.position = 0.0
		rightEncoder.position = 0.0
	}
	
	fun resetGyro() = navX.reset()
	
	override fun initDefaultCommand() {
		defaultCommand = Teleop()
	}
	
	override fun initSendable(builder: SendableBuilder) {
		builder.setSmartDashboardType("Drives")
		builder.setActuator(true)
		builder.setSafeState { speed = 0.0 }
		builder.addDoubleProperty("leftDistance", ::leftDistance, null)
		builder.addDoubleProperty("rightDistance", ::rightDistance, null)
		val firmwareSupplier = { arrayOf<String>("Left Front " + leftFrontSpark.firmwareString, "Left Back " + leftBackSpark.getFirmwareString(), "Right Front" + rightFrontSpark.getFirmwareString(), "Right Back " + rightBackSpark.getFirmwareString()) }
		builder.addStringArrayProperty("firmware", firmwareSupplier, null)
		builder.addDoubleProperty("Gyro", ::gyroAngle, null)
	}
	
	override fun pidWrite(output: Double) = setSpeed(output, output)
	
}
